src/streams.cpp

```diff
@@ -23,8 +25,9 @@ std::size_t AutoFile::detail_fread(Span<std::byte> dst)
 
 void AutoFile::read(Span<std::byte> dst)
 {
-    if (detail_fread(dst) != dst.size()) {
-        throw std::ios_base::failure(feof() ? "AutoFile::read: end of file" : "AutoFile::read: fread failed");
+    if (detail_fread(dst) != dst.size() || ferror(m_file)) {
```

Comment:
This was somewhat answered in https://github.com/bitcoin/bitcoin/pull/29307#discussion_r1465152180 above but let me elaborate - the thing that prompted me to do this change is this text from the `fread(3)` manual page:

> The function fread() does not distinguish between end-of-file and error,
> and callers must use feof(3) and ferror(3) to determine which occurred.

I think checking for `ferror(3)` after `fread(3)` is a kind of harmless "belt-and-suspenders". If you are sure that if `detail_fread(dst) == dst.size()` then `ferror(m_file)` will always be `false` or feel strongly against this then I could drop the commit de23848eed `util: check for error after reading from a file`.

Reply:
Well, it is pretty much obvious that checking for an error after read is either:

* a noop, if the read could have never failed because it managed to read as many bytes as requested or
* a bug fix where a read error was ignored before the fix

so this is a safe change (it does not break anything), possibly fixing a bug. Given that you keep pushing against this I will drop this commit in the hopes to move this PR forward.

Reply:
I just found a similar code to this proposed change:

https://github.com/bitcoin/bitcoin/blob/2d21060af831fa8f8f1791b4464961d5f28a88cb/src/node/utxo_snapshot.cpp#L74-L80

After reading from a file and getting everything that is needed from it, the code checks if there is an extra data and if that read fails it reports an error.

Reply:
I think I see your point, you prefer 1. over 2., assuming we got all the bytes we asked for from a read, then:

1. Ignore possible IO error by not checking for it
2. Check for an error and behave as if the read failed

I dropped this change.